extruct/_extruct.py

```diff
@@ -38,48 +40,109 @@ def extract(htmlstring, base_url=None, encoding="UTF-8",
+            DeprecationWarning,
+            stacklevel=2)
         base_url = kwargs.pop('url')
     if kwargs:
         raise TypeError('Unexpected keyword arguments')
-    if not (isinstance(syntaxes, list) and all(v in SYNTAXES for v in syntaxes)):
+    if not (isinstance(syntaxes, list) and all(v in SYNTAXES
+                                               for v in syntaxes)):
         raise ValueError("syntaxes must be a list with any or all (default) of"
                          "these values: {}".format(SYNTAXES))
     if errors not in ['log', 'ignore', 'strict']:
         raise ValueError('Invalid error command, valid values are either "log"'
                          ', "ignore" or "strict"')
-    tree = parse_xmldom_html(htmlstring, encoding=encoding)
+    try:
+        tree = parse_xmldom_html(htmlstring, encoding=encoding)
+    except Exception as e:
+        if errors == 'ignore':
+            return {}
+        if errors == 'log':
+            logger.exception(
+                f'Failed to parse html, raises {e}')
+            return {}
+        if errors == 'strict':
+            raise e
     processors = []
     if 'microdata' in syntaxes:
-        processors.append(('microdata', MicrodataExtractor(add_html_node=return_html_node).extract_items, tree))
+        processors.append(
+            ('microdata',
+             MicrodataExtractor(add_html_node=return_html_node).extract_items,
+             tree
+             ))
     if 'json-ld' in syntaxes:
-        processors.append(('json-ld', JsonLdExtractor().extract_items, tree))
+        processors.append(
+            ('json-ld',
+             JsonLdExtractor().extract_items,
+             tree,
+             ))
     if 'opengraph' in syntaxes:
-        processors.append(('opengraph', OpenGraphExtractor().extract_items, tree))
+        processors.append(
+            ('opengraph',
+             OpenGraphExtractor().extract_items,
+             tree
+             ))
     if 'microformat' in syntaxes:
-        processors.append(('microformat', MicroformatExtractor().extract_items, htmlstring))
+        processors.append(
+            ('microformat',
+             MicroformatExtractor().extract_items,
+             htmlstring
+             ))
     if 'rdfa' in syntaxes:
-        processors.append(('rdfa', RDFaExtractor().extract_items, tree))
+        processors.append(
+            ('rdfa', RDFaExtractor().extract_items,
+             tree,
+             ))
     output = {}
-    for label, extract, document in processors:
+    for syntax, extract, document in processors:
         try:
-            output[label] = list(extract(document, base_url=base_url))
-        except Exception:
+            output[syntax] = list(extract(document, base_url=base_url))
+        except Exception as e:
             if errors == 'log':
-                logger.exception('Failed to extract {}'.format(label))
+                logger.exception(f'Failed to extract {syntax}, raises {e}')
```

Comment:
f syntax appeared only in python 3.6, we'll have to use an older syntax for python 2.7+ 😢

Reply:
ah, darn! It makes the code look much neater!